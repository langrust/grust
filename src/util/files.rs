use codespan_reporting::files;
use std::ops::Range;

/// A file that is backed by an `Arc<String>`.
#[derive(Debug, Clone)]
struct File {
    /// The name of the file.
    name: String,
    /// The source code of the file.
    source: String,
    /// The starting byte indices in the source code.
    line_starts: Vec<usize>,
}

impl File {
    fn line_start(&self, line_index: usize) -> Result<usize, files::Error> {
        use std::cmp::Ordering;

        match line_index.cmp(&self.line_starts.len()) {
            Ordering::Less => Ok(*self
                .line_starts
                .get(line_index)
                .expect("failed despite previous check")),
            Ordering::Equal => Ok(self.source.len()),
            Ordering::Greater => Err(files::Error::LineTooLarge {
                given: line_index,
                max: self.line_starts.len() - 1,
            }),
        }
    }
}

/// An opaque file identifier.
#[derive(Copy, Clone, PartialEq, Eq, Debug)]
pub struct FileId(i32);
impl FileId {
    /// Construct and return the default file identifier.
    pub fn default() -> Self {
        return FileId(-1);
    }
}

/// The files handler.
#[derive(Debug, Clone)]
pub struct Files {
    files: Vec<File>,
}

impl Files {
    /// Create a new files database.
    pub fn new() -> Files {
        Files { files: Vec::new() }
    }

    /// Add a file to the database, returning the handle that can be used to
    /// refer to it again.
    pub fn add(
        &mut self,
        name: impl Into<String>,
        source: impl Into<String>,
    ) -> Option<FileId> {
        let file_id = FileId(i32::try_from(self.files.len()).ok()?);
        let name = name.into();
        let source = source.into();
        let line_starts = files::line_starts(&source).collect();

        self.files.push(File {
            name,
            line_starts,
            source,
        });

        Some(file_id)
    }

    /// Get the file corresponding to the given id.
    fn get(&self, file_id: FileId) -> Result<&File, files::Error> {
        self.files
            .get(file_id.0 as usize)
            .ok_or(files::Error::FileMissing)
    }
}

impl<'files> files::Files<'files> for Files {
    type FileId = FileId;
    type Name = &'files str;
    type Source = &'files str;

    fn name(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(self.get(file_id)?.name.as_ref())
    }

    fn source(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(&self.get(file_id)?.source)
    }

    fn line_index(
        &self,
        file_id: FileId,
        byte_index: usize
    ) -> Result<usize, files::Error> {
        self.get(file_id)?
            .line_starts
            .binary_search(&byte_index)
            .or_else(|next_line| Ok(next_line - 1))
    }

    fn line_range(
        &self,
        file_id: FileId,
        line_index: usize,
    ) -> Result<Range<usize>, files::Error> {
        let file = self.get(file_id)?;
        let line_start = file.line_start(line_index)?;
        let next_line_start = file.line_start(line_index + 1)?;

        Ok(line_start..next_line_start)
    }
}