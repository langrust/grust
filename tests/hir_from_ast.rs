use codespan_reporting::files::SimpleFiles;

use grustine::{error::display, hir_from_ast};

#[test]
fn hir_from_ast_transformation_for_counter() {
    let mut files = SimpleFiles::new();

    let counter_id = files.add(
        "counter.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/success/counter.gr")
            .expect("unkown file"),
    );

    match hir_from_ast(counter_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_blinking() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "blinking.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/success/blinking.gr")
            .expect("unkown file"),
    );

    match hir_from_ast(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_button_management() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "button_management.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/success/button_management.gr")
            .expect("unkown file"),
    );

    match hir_from_ast(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_button_management_condition_match() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "button_management_condition_match.gr",
        std::fs::read_to_string(
            "tests/fixture/hir_from_ast/success/button_management_condition_match.gr",
        )
        .expect("unkown file"),
    );

    match hir_from_ast(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_button_management_using_function() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "button_management_using_function.gr",
        std::fs::read_to_string(
            "tests/fixture/hir_from_ast/success/button_management_using_function.gr",
        )
        .expect("unkown file"),
    );

    match hir_from_ast(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_pid() {
    let mut files = SimpleFiles::new();

    let pid_id = files.add(
        "pid.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/success/pid.gr").expect("unkown file"),
    );

    match hir_from_ast(pid_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn error_when_hir_from_ast_transformation_for_counter_unknown_signal() {
    let mut files = SimpleFiles::new();

    let counter_unknown_signal_id = files.add(
        "counter_unknown_signal.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/error/counter_unknown_signal.gr")
            .expect("unkown file"),
    );

    match hir_from_ast(counter_unknown_signal_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn error_when_hir_from_ast_transformation_for_button_management_using_function_unknown_element() {
    let mut files = SimpleFiles::new();

    let button_management_using_function_unknown_element_id = files.add(
        "button_management_using_function_unknown_element.gr",
        std::fs::read_to_string(
            "tests/fixture/hir_from_ast/error/button_management_using_function_unknown_element.gr",
        )
        .expect("unkown file"),
    );

    match hir_from_ast(
        button_management_using_function_unknown_element_id,
        &mut files,
    ) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn error_when_hir_from_ast_transformation_for_button_management_unknown_type() {
    let mut files = SimpleFiles::new();

    let button_management_unknown_type_id = files.add(
        "button_management_unknown_type.gr",
        std::fs::read_to_string(
            "tests/fixture/hir_from_ast/error/button_management_unknown_type.gr",
        )
        .expect("unkown file"),
    );

    match hir_from_ast(button_management_unknown_type_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn error_when_hir_from_ast_transformation_for_pid_unknown_field() {
    let mut files = SimpleFiles::new();

    let pid_unknown_field_id = files.add(
        "pid_unknown_field.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/error/pid_unknown_field.gr")
            .expect("unkown file"),
    );

    match hir_from_ast(pid_unknown_field_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn error_when_typing_pid_missing_field() {
    let mut files = SimpleFiles::new();

    let pid_missing_field_id = files.add(
        "pid_missing_field.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/error/pid_missing_field.gr")
            .expect("unkown file"),
    );

    match hir_from_ast(pid_missing_field_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn error_when_typing_blinking_component_call() {
    let mut files = SimpleFiles::new();

    let blinking_component_call_id = files.add(
        "blinking_component_call.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/error/blinking_component_call.gr")
            .expect("unkown file"),
    );

    match hir_from_ast(blinking_component_call_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn error_when_typing_blinking_already_defined_element() {
    let mut files = SimpleFiles::new();

    let blinking_already_defined_element_id = files.add(
        "blinking_already_defined_element.gr",
        std::fs::read_to_string(
            "tests/fixture/hir_from_ast/error/blinking_already_defined_element.gr",
        )
        .expect("unkown file"),
    );

    match hir_from_ast(blinking_already_defined_element_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn error_when_typing_pid_expect_structure_type() {
    let mut files = SimpleFiles::new();

    let pid_expect_structure_type_id = files.add(
        "pid_expect_structure_type.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/error/pid_expect_structure_type.gr")
            .expect("unkown file"),
    );

    match hir_from_ast(pid_expect_structure_type_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn error_when_typing_pid_unknown_enumeration() {
    let mut files = SimpleFiles::new();

    let pid_unknown_enumeration_id = files.add(
        "pid_unknown_enumeration.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/error/pid_unknown_enumeration.gr")
            .expect("unkown file"),
    );

    match hir_from_ast(pid_unknown_enumeration_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_urban_braking() {
    let mut files = SimpleFiles::new();

    let urban_braking_id = files.add(
        "urban_braking.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/success/urban_braking.gr").expect("unkown file"),
    );

    match hir_from_ast(urban_braking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}
