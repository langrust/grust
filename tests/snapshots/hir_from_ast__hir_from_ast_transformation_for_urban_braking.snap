---
source: tests/hir_from_ast.rs
expression: file
---
typedefs:
  - id: 20
    kind:
      Enumeration:
        elements:
          - 17
          - 18
          - 19
    location: ~
functions:
  - id: 23
    statements:
      - id: 29
        expression:
          kind:
            Application:
              function_expression:
                kind:
                  Identifier:
                    id: 3
                typing: ~
                location: ~
                dependencies: ~
              inputs:
                - kind:
                    Identifier:
                      id: 22
                  typing: ~
                  location: ~
                  dependencies: ~
                - kind:
                    Constant:
                      constant:
                        Float: 0.278
                  typing: ~
                  location: ~
                  dependencies: ~
          typing: ~
          location: ~
          dependencies: ~
        location: ~
    returned:
      kind:
        Application:
          function_expression:
            kind:
              Identifier:
                id: 15
            typing: ~
            location: ~
            dependencies: ~
          inputs:
            - kind:
                Application:
                  function_expression:
                    kind:
                      Identifier:
                        id: 14
                    typing: ~
                    location: ~
                    dependencies: ~
                  inputs:
                    - kind:
                        Identifier:
                          id: 29
                      typing: ~
                      location: ~
                      dependencies: ~
                    - kind:
                        Identifier:
                          id: 21
                      typing: ~
                      location: ~
                      dependencies: ~
              typing: ~
              location: ~
              dependencies: ~
            - kind:
                Enumeration:
                  enum_id: 20
                  elem_id: 18
              typing: ~
              location: ~
              dependencies: ~
            - kind:
                Enumeration:
                  enum_id: 20
                  elem_id: 19
              typing: ~
              location: ~
              dependencies: ~
      typing: ~
      location: ~
      dependencies: ~
    location: ~
nodes: []
component:
  id: 28
  unscheduled_equations:
    26:
      id: 26
      expression:
        kind:
          Expression:
            expression:
              When:
                id: 30
                option:
                  kind:
                    Expression:
                      expression:
                        Identifier:
                          id: 24
                  typing: ~
                  location: ~
                  dependencies: ~
                present:
                  kind:
                    Expression:
                      expression:
                        Application:
                          function_expression:
                            kind:
                              Expression:
                                expression:
                                  Identifier:
                                    id: 23
                            typing: ~
                            location: ~
                            dependencies: ~
                          inputs:
                            - kind:
                                Expression:
                                  expression:
                                    Identifier:
                                      id: 30
                              typing: ~
                              location: ~
                              dependencies: ~
                            - kind:
                                Expression:
                                  expression:
                                    Identifier:
                                      id: 25
                              typing: ~
                              location: ~
                              dependencies: ~
                  typing: ~
                  location: ~
                  dependencies: ~
                present_body: []
                default:
                  kind:
                    Expression:
                      expression:
                        Identifier:
                          id: 27
                  typing: ~
                  location: ~
                  dependencies: ~
                default_body: []
        typing: ~
        location: ~
        dependencies: ~
      location: ~
    27:
      id: 27
      expression:
        kind:
          FollowedBy:
            constant:
              kind:
                Expression:
                  expression:
                    Enumeration:
                      enum_id: 20
                      elem_id: 17
              typing: ~
              location: ~
              dependencies: ~
            expression:
              kind:
                Expression:
                  expression:
                    Identifier:
                      id: 26
              typing: ~
              location: ~
              dependencies: ~
        typing: ~
        location: ~
        dependencies: ~
      location: ~
  unitary_nodes: {}
  contract:
    requires:
      - kind:
          Binary:
            op: And
            left:
              kind:
                Binary:
                  op: Leq
                  left:
                    kind:
                      Constant:
                        constant:
                          Float: 0
                    location: ~
                  right:
                    kind:
                      Identifier:
                        id: 25
                    location: ~
              location: ~
            right:
              kind:
                Binary:
                  op: Low
                  left:
                    kind:
                      Identifier:
                        id: 25
                    location: ~
                  right:
                    kind:
                      Constant:
                        constant:
                          Float: 50
                    location: ~
              location: ~
        location: ~
    ensures: []
    invariant: []
  location: ~
  graph: {}
interface:
  id: 36
  flow_statements:
    - id: 34
      flow_expression:
        kind:
          Timeout:
            flow_expression:
              kind:
                Merge:
                  flow_expression_1:
                    kind:
                      Ident:
                        id: 32
                    typing: ~
                    location: ~
                  flow_expression_2:
                    kind:
                      Ident:
                        id: 33
                    typing: ~
                    location: ~
              typing: ~
              location: ~
            timeout_ms: 500
        typing: ~
        location: ~
    - id: 35
      flow_expression:
        kind:
          ComponentCall:
            component_id: 28
            inputs:
              - - 24
                - kind:
                    Ident:
                      id: 34
                  typing: ~
                  location: ~
              - - 25
                - kind:
                    Ident:
                      id: 31
                  typing: ~
                  location: ~
            signal_id: 26
        typing: ~
        location: ~
location: ~
